beyond-all-reason/Beyond-All-Reason
# BAR Hotkeys: UIkeys keybind format, deprecate .lua keybind format

This is a continuation of https://github.com/beyond-all-reason/Beyond-All-Reason/pull/1403 from @badosu 

> With the addition of keyload/keyreload on uikeys with arguments and
keysave with argument now we can safely switch back to .txt and make bar
hotkeys finally compatible with uikeys.txt

> Users that previously had to disable bar hotkeys now can simply use
their uikeys with the preset custom enabled.

> To load engine defaults simply put add keydefaults to the top of the file.

> Users of custom preset on older format seamlessly migrate to new format, being written to uikeys_bar.txt.

### Overview
This deprecates the `.lua` keybind files, and adds automatic conversion to the older and more fully-featured and easier to edit `uikeys.txt` format. This will bring us back to just having a single keybind format, and with the addition of `keyload` users will be able to *extend* existing keybind presets instead of fully re-declaring them. This will make the process of creating custom keybinds much less error-prone and will allow users to stay up-to-date with changes to presets without having to edit their custom keybinds (unless there is a conflict). No longer will users have to disable the `BAR Hotkeys` widget.

### To test
- Please back up any custom keybinds you already have! I don't want you to lose anything if there are errors in this branch
- Load this PR with a preset already enabled, and expect to have functional keybinds
- Load this PR with a *custom* `bar_hotkeys_custom.lua` file, and expect it to be accurately converted into a `uikeys.txt` file
- Load this PR with a *custom* preset with a *non-qwerty* keyboard layout, and expect it to be accurately converted into a `uikeys.txt` file.
- Switch presets and expect everything to update seamlessly
- Expect `uikeys_backup.txt` to be created when the game launches
